jobmaster/cvsJob/cvsJobMaster.go

```diff
@@ -82,87 +87,131 @@ func NewCVSJobMaster(ctx *dcontext.Context, workerID lib.WorkerID, masterID lib.
+	// This persistence has to succeed before we set this master to normal status.
+	statusBytes, err := json.Marshal(jm.jobStatus)
+	if err != nil {
+		return err
+	}
+	err = jm.UserMetaKVClient().Store(ctx, jm.workerID, string(statusBytes))
+	if err != nil {
+		return err
+	}
+	jm.setStatusCode(lib.WorkerStatusNormal)
 	return nil
 }
 
 func (jm *JobMaster) Tick(ctx context.Context) error {
 	filesNum := 0
 	jm.counter = 0
-	if jm.getStatusCode() == lib.WorkerStatusInit {
-		err := jm.UpdateJobStatus(ctx, jm.Status())
-		if err != nil {
-			log.L().Warn("update job status failed", zap.Any("id", jm.workerID), zap.Error(err))
-		} else {
-			jm.setStatusCode(lib.WorkerStatusNormal)
-		}
+	if !jm.IsMasterReady() {
+		log.L().Info("jobmaster is not ready", zap.Any("master id", jm.workerID))
+		return nil
 	}
-	jm.syncFilesInfoMu.Lock()
-	defer jm.syncFilesInfoMu.Unlock()
-	for _, worker := range jm.syncFilesInfo {
-		if worker.handle == nil {
+
+	for idx, workerInfo := range jm.syncFilesInfo {
+		// check if need to recreate worker
+		if workerInfo.needCreate.Load() {
+			workerID, err := jm.CreateWorker(lib.CvsTask, getTaskConfig(jm.jobStatus, idx), 10)
+			if err != nil {
+				log.L().Warn("create worker failed, try next time", zap.Any("master id", jm.workerID), zap.Error(err))
+			} else {
+				jm.launchedWorkers.Store(workerID, idx)
+				workerInfo.needCreate.Store(false)
+			}
 			continue
 		}
-		status := worker.handle.Status()
+
+		// still awaiting online
+		if workerInfo.handle.Load() == nil {
+			continue
+		}
+
+		// update job status
+		handle := *(*lib.WorkerHandle)(workerInfo.handle.Load())
+		status := handle.Status()
 		switch status.Code {
 		case lib.WorkerStatusNormal, lib.WorkerStatusFinished, lib.WorkerStatusStopped:
-			num, err := strconv.ParseInt(string(status.ExtBytes), 10, 64)
+			taskStatus := &cvsTask.Status{}
+			err := json.Unmarshal(status.ExtBytes, taskStatus)
 			if err != nil {
 				return err
 			}
-			worker.curLoc = num
-			jm.counter += num
-			log.L().Debug("cvs job tmp num ", zap.Any("id", worker.handle.ID()), zap.Int64("counter", num), zap.Any("status", status.Code))
+
+			jm.jobStatus.FileInfos[idx].Location = taskStatus.CurrentLoc
+			jm.counter += taskStatus.Count
+
+			log.L().Debug("cvs job tmp num ", zap.Any("id", idx), zap.Any("status", string(status.ExtBytes)))
 			// todo : store the sync progress into the meta store for each file
 			if status.Code == lib.WorkerStatusFinished {
 				filesNum++
 			}
 		case lib.WorkerStatusError:
-			log.L().Error("sync file failed ", zap.Any("idx", worker.idx))
+			log.L().Error("sync file failed ", zap.Any("idx", idx))
 		default:
 			log.L().Info("worker status abnormal", zap.Any("status", status))
 		}
 	}
 	if jm.statusRateLimiter.Allow() {
+		statsBytes, err := json.Marshal(jm.jobStatus)
+		if err != nil {
+			log.L().Warn("serialize job status failed, try next time", zap.Any("master id", jm.workerID), zap.Error(err))
```

Comment:
Should return immediately here?

Reply:
Good Point, It's disaster if we cannot encode successfully.

Reply:
done, PTAL